cmd/createorupdate/update.go

```diff
@@ -2,27 +2,84 @@ package main
 
 import (
 	"context"
+	"reflect"
 
 	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2018-06-01/compute"
 	"github.com/Azure/go-autorest/autorest/azure/auth"
 	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/go-test/deep"
 
 	"github.com/openshift/openshift-azure/pkg/api"
 	"github.com/openshift/openshift-azure/pkg/log"
 )
 
-func update(ctx context.Context, cs *api.OpenShiftManagedCluster, p api.Plugin) error {
+func update(ctx context.Context, cs, oldCs *api.OpenShiftManagedCluster, p api.Plugin, azuredeploy []byte) error {
 	config := auth.NewClientCredentialsConfig(ctx.Value(api.ContextKeyClientID).(string), ctx.Value(api.ContextKeyClientSecret).(string), ctx.Value(api.ContextKeyTenantID).(string))
 	authorizer, err := config.Authorizer()
 	if err != nil {
 		return err
 	}
-
 	ssc := compute.NewVirtualMachineScaleSetsClient(cs.Properties.AzProfile.SubscriptionID)
 	ssc.Authorizer = authorizer
 	vmc := compute.NewVirtualMachineScaleSetVMsClient(cs.Properties.AzProfile.SubscriptionID)
 	vmc.Authorizer = authorizer
 
+	// Need to determine whether the current update is a scale operation before
+	// applying the ARM template. For scale up, we need to figure out which are
+	// the new VMs that were created and wait for them to turn ready. For scale
+	// down, we need to drain the correct VMs before applying the ARM template
+	// scales them down.
+	isScaleOp := isScaleUpdate(cs, oldCs)
+	vmsBefore := map[string]struct{}{}
+	if isScaleOp {
+		for _, agent := range cs.Properties.AgentPoolProfiles {
+			vms, err := listVMs(ctx, cs, vmc, agent.Role)
+			if err != nil {
+				return err
+			}
+
+			if len(vms) > agent.Count {
+				for _, vm := range vms[agent.Count:] {
+					if err := drain(ctx, cs, p, vmc, agent.Role, *vm.InstanceID, *vm.VirtualMachineScaleSetVMProperties.OsProfile.ComputerName); err != nil {
+						return err
+					}
+				}
+			} else {
+				for _, vm := range vms {
+					vmsBefore[*vm.VirtualMachineScaleSetVMProperties.OsProfile.ComputerName] = struct{}{}
+				}
+			}
+		}
+	}
+
+	// Apply the ARM template
+	if err := deploy(ctx, cs, p, azuredeploy); err != nil {
+		return err
+	}
+
+	if isScaleOp {
+		for _, agent := range cs.Properties.AgentPoolProfiles {
+			vms, err := listVMs(ctx, cs, vmc, agent.Role)
```

Comment:
Another thing we avoided with `isScaleOp` was listing after deploy when we are scaling down.